### PR TITLE
Sentry chart

### PR DESCRIPTION
**Pull request by rothgar**
My first chart. Still in progress. Feedback welcome.

### CODE REVIEW

**Comment by rothgar** [comment]
Trying to install for testing. Ran

```
helm dep build
helm install --debug .
```

Seeing errors
```
Created tunnel using local port: '36815'
SERVER: "localhost:36815"
CHART PATH: /home/jgarr/src/kubernetes/charts/incubator/sentry
Error: release xrayed-terrier failed: deployments.extensions "xrayed-terrier-sentry" already exists
```
This is a brand new GKE cluster for testing
```
$ helm version
Client: &version.Version{SemVer:"v2.2.3", GitCommit:"1402a4d6ec9fb349e17b912e32fe259ca21181e3", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.2.3", GitCommit:"1402a4d6ec9fb349e17b912e32fe259ca21181e3", GitTreeState:"clean"}
```

Any thoughts?
@lachie83

File: incubator/sentry/values.yaml
Comment (lachie83) [inline incubator/sentry/values.yaml]: Remove indentation on persistence to top level key.

**Comment by lachie83** [comment]
@rothgar This PR is great to see. This issue is because both the web, worker and cron deployments are being generated with the same name. Try appending the component name to the end. Here's an example -- https://github.com/kubernetes/charts/blob/master/incubator/istio/templates/manager-deployment.yaml#L5. You could also do something like `name: {{ template "fullname" . }}-worker`

**Comment by lachie83** [comment]
I got this chart to install with the following changes

```
~/sandbox/kubernetes/charts/incubator/sentry on pr/1226 [!]
$ git d
 incubator/sentry/requirements.lock                 |  6 ++++--
 incubator/sentry/templates/cron-deployment.yaml    |  2 +-
 incubator/sentry/templates/workers-deployment.yaml |  2 +-
 incubator/sentry/values.yaml                       | 20 ++++++++++----------
 4 files changed, 16 insertions(+), 14 deletions(-)

diff --git a/incubator/sentry/requirements.lock b/incubator/sentry/requirements.lock
index 581a112..5679370 100644
--- a/incubator/sentry/requirements.lock
+++ b/incubator/sentry/requirements.lock
@@ -1,15 +1,17 @@
 dependencies:
 - condition: ""
   enabled: false
+  import-values: null
   name: postgresql
   repository: https://kubernetes-charts.storage.googleapis.com/
   tags: null
   version: 0.7.1
 - condition: ""
   enabled: false
+  import-values: null
   name: redis
   repository: https://kubernetes-charts.storage.googleapis.com/
   tags: null
   version: 0.6.0
-digest: sha256:ea6510a357c699082e455250d166978c363f420ee8d589c48492e83b4c376eba
-generated: 2017-06-02T06:18:46.211627938Z
+digest: sha256:dbaa80dbd8326dc559176f3bb5c97723b5a9c6378cd5b01fe3d0787825de9bac
+generated: 2017-06-02T10:11:05.87979944-07:00
diff --git a/incubator/sentry/templates/cron-deployment.yaml b/incubator/sentry/templates/cron-deployment.yaml
index ae212fc..94b9c25 100644
--- a/incubator/sentry/templates/cron-deployment.yaml
+++ b/incubator/sentry/templates/cron-deployment.yaml
@@ -1,7 +1,7 @@
 apiVersion: extensions/v1beta1
 kind: Deployment
 metadata:
-  name: {{ template "fullname" . }}
+  name: {{ template "fullname" . }}-cron
   labels:
     chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
 spec:
diff --git a/incubator/sentry/templates/workers-deployment.yaml b/incubator/sentry/templates/workers-deployment.yaml
index 374004b..ed1862b 100644
--- a/incubator/sentry/templates/workers-deployment.yaml
+++ b/incubator/sentry/templates/workers-deployment.yaml
@@ -1,7 +1,7 @@
 apiVersion: extensions/v1beta1
 kind: Deployment
 metadata:
-  name: {{ template "fullname" . }}
+  name: {{ template "fullname" . }}-worker
   labels:
     chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
 spec:
diff --git a/incubator/sentry/values.yaml b/incubator/sentry/values.yaml
index 6006b51..08b929d 100644
--- a/incubator/sentry/values.yaml
+++ b/incubator/sentry/values.yaml
@@ -21,17 +21,17 @@ resources:
     cpu: 100m
     memory: 128Mi

-  ## Enable persistence using Persistent Volume Claims
-  ## ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
+## Enable persistence using Persistent Volume Claims
+## ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
+##
+persistence:
+  enabled: false
+  ## If defined, volume.beta.kubernetes.io/storage-class: <storageClass>
+  ## Default: volume.alpha.kubernetes.io/storage-class: default
   ##
-  persistence:
-    enabled: false
-    ## If defined, volume.beta.kubernetes.io/storage-class: <storageClass>
-    ## Default: volume.alpha.kubernetes.io/storage-class: default
-    ##
-    # storageClass:
-    accessMode: ReadWriteOnce
-    size: 8Gi
+  # storageClass:
+  accessMode: ReadWriteOnce
+  size: 8Gi

 ## Configure ingress resource that allow you to access the
 ## Sentry instalation. Set up the URL
```

```
~/sandbox/kubernetes/charts/incubator/sentry on pr/1226 [!]
$ helm install .
NAME:   agile-gibbon
LAST DEPLOYED: Fri Jun  2 10:21:44 2017
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/PersistentVolumeClaim
NAME                     STATUS  VOLUME                                    CAPACITY  ACCESSMODES  STORAGECLASS  AGE
agile-gibbon-postgresql  Bound   pvc-f4c0d081-47b7-11e7-ae3f-000d3a72314b  8Gi       RWO          7s

==> v1/Service
NAME                     CLUSTER-IP    EXTERNAL-IP  PORT(S)         AGE
agile-gibbon-redis       10.0.160.136  <none>       6379/TCP        6s
agile-gibbon-sentry      10.0.198.236  <pending>    9000:31719/TCP  5s
agile-gibbon-postgresql  10.0.236.82   <none>       5432/TCP        5s

==> v1beta1/Deployment
NAME                        DESIRED  CURRENT  UP-TO-DATE  AVAILABLE  AGE
agile-gibbon-sentry         1        0        0           0          4s
agile-gibbon-sentry-worker  2        0        0           0          4s
agile-gibbon-redis          1        1        1           0          4s
agile-gibbon-postgresql     1        1        1           0          4s
agile-gibbon-sentry-cron    1        0        0           0          4s

==> v1/Secret
NAME                     TYPE    DATA  AGE
agile-gibbon-sentry      Opaque  2     7s
agile-gibbon-postgresql  Opaque  1     7s
agile-gibbon-redis       Opaque  1     7s


NOTES:
1. Get the application URL by running these commands:
     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
           You can watch the status of by running 'kubectl get svc -w agile-gibbon-sentry'
  export SERVICE_IP=$(kubectl get svc --namespace default agile-gibbon-sentry -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
  echo http://$SERVICE_IP:9000
```

**Comment by rothgar** [comment]
I have a job that needs to run once during installation. Is there an example I can look at for this? For a fresh install I need to initialize the database with `sentry upgrade`

File: incubator/sentry/templates/cron-deployment.yaml
Comment (rothgar) [inline incubator/sentry/templates/cron-deployment.yaml]: This isn't getting set in the pod? I manually set it in values but it's not being pulled in. Any ideas?

Comment (ritazh) [reply incubator/sentry/templates/cron-deployment.yaml]: Maybe try [this](https://github.com/kubernetes/charts/blob/master/stable/gitlab-ee/templates/deployment.yaml#L47-L51) in deployment.yaml and [this](https://github.com/kubernetes/charts/blob/master/stable/gitlab-ee/templates/secrets.yaml#L17) in secret.yaml and [this](https://github.com/kubernetes/charts/blob/master/stable/gitlab-ee/values.yaml#L83) in values.yaml.

**Comment by rothgar** [comment]
I found the hooks docs (examples also in spinnaker). Added db init and user creation.

Still need

- [ ] smtp settings
- [ ] persistant storage
- [ ] [plugins](https://docs.sentry.io/server/plugins/)

**Comment by rothgar** [comment]
Just for reference, I'm also running into this issue kubernetes/helm#1769 with the jobs

**Comment by rothgar** [comment]
@lachie83 have an example of disabling sub charts if they're not needed (eg postgres)? I found this https://docs.helm.sh/chart-best-practices/#conditions-and-tags but not sure how to implement that for a subchart. I searched the repo (not very well) and didn't see an example.

**Comment by lachie83** [comment]
Unfortunately you cannot use the conditions on local subcharts only with dependencies as you mentioned above. One way I've see before is to set a global flag in the top chart values.yaml and conditionalize creation of all resources in the subchart on that flag. eg

```
{{- if .Values.global.postgres.enabled }}
{{ end }}
```

Does that help?

**Comment by rothgar** [comment]
I think this is at a minimal viable chart.

Can someone test in a different environment for me?

**Comment by lachie83** [comment]
All pods come up and running and the job exits success following an install. How else would you like me to test? Is there something functional I can do?
```
bailing-possum-postgresql-2744245312-nwv54       1/1       Running   0          6m
bailing-possum-redis-46185774-1w33d              1/1       Running   0          6m
bailing-possum-sentry-884421309-5k861            1/1       Running   0          6m
bailing-possum-sentry-cron-3640742245-4s8pb      1/1       Running   1          6m
bailing-possum-sentry-worker-1644057744-8hwzq    1/1       Running   1          6m
bailing-possum-sentry-worker-1644057744-lfqvd    1/1       Running   1          6m
kubectl get jobs
NAME                     DESIRED   SUCCESSFUL   AGE
bailing-possum-db-init   1         1            7m
```

**Comment by lachie83** [comment]
Actually it looks like the sentry pod keeps crashing

```
bailing-possum-sentry-884421309-5k861            0/1       CrashLoopBackOff   4          11m
```

Here are the logs
```
*** Starting uWSGI 2.0.15 (64bit) on [Thu Jun  8 06:58:23 2017] ***
compiled with version: 4.9.2 on 02 June 2017 13:28:26
os: Linux-4.4.0-77-generic #98-Ubuntu SMP Wed Apr 26 08:34:02 UTC 2017
nodename: bailing-possum-sentry-884421309-5k861
machine: x86_64
clock source: unix
detected number of CPU cores: 2
current working directory: /
detected binary path: /usr/local/bin/uwsgi
!!! no internal routing support, rebuild with pcre support !!!
your memory page size is 4096 bytes
detected max file descriptor number: 65536
lock engine: pthread robust mutexes
thunder lock: enabled
uwsgi socket 0 bound to TCP address 0.0.0.0:9000 fd 3
Python version: 2.7.13 (default, May 11 2017, 21:39:56)  [GCC 4.9.2]
Set PythonHome to /usr/local
Python main interpreter initialized at 0xd64260
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
setting request body buffering size to 65536 bytes
mapped 1922048 bytes (1877 KB) for 12 cores
*** Operational MODE: preforking+threaded ***
spawned uWSGI master process (pid: 11)
spawned uWSGI worker 1 (pid: 16, cores: 4)
spawned uWSGI worker 2 (pid: 17, cores: 4)
spawned uWSGI worker 3 (pid: 18, cores: 4)

```

File: incubator/sentry/values.yaml
Comment (gianrubio) [inline incubator/sentry/values.yaml]: This limits are too low for sentry starting. I recommend at least 300m for cpu and 300Mi for memory.

File: incubator/sentry/templates/web-deployment.yaml
Comment (gianrubio) [inline incubator/sentry/templates/web-deployment.yaml]: Please add probes (this are my current probes)

```
          livenessProbe:
            failureThreshold: 5
            httpGet:
              path: /_health/
              port: 9000
              scheme: HTTP
            initialDelaySeconds: 50
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 2         
          readinessProbe:
            failureThreshold: 10
            httpGet:
              path: /_health/
              port: 9000
              scheme: HTTP
            initialDelaySeconds: 50
            periodSeconds: 10
            successThreshold: 1
            timeoutSeconds: 2
```

**Comment by gianrubio** [comment]
> Actually it looks like the sentry pod keeps crashing

@lachie83 please take  a look on my comments about the resources. This values are too low to sentry start the application.

File: incubator/sentry/templates/web-deployment.yaml
Comment (rothgar) [reply incubator/sentry/templates/web-deployment.yaml]: Thanks, I initially tried using the log in page as a health check but removed it because I didn't look up what endpoint should be checked instead. I'll add this along with better limits.

**Comment by rothgar** [comment]
@lachie83 I think I'm mostly curious about what are requirements for the chart to be merged into the main repo? Are there typical tests that are performed? Are charts checked against k8s clusters in GKE, AWS, Azure?

I'm working on the changes @gianrubio suggested but after that I'm not sure how much else needs to be done before it can be merged.

**Comment by lachie83** [comment]
@rothgar here's the technical checklist on getting it merged

https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

The current charts are tested against a 1.6 cluster running on GKE only. They only test lint and installation through to a running pod. You could also add helm tests which we hope to use eventually. See https://github.com/kubernetes/charts/blob/master/stable/consul/templates/test-config.yaml for example

File: incubator/sentry/Chart.yaml
Comment (lachie83) [inline incubator/sentry/Chart.yaml]: Add maintainers please use your github name under the name section and an email address

**Comment by lachie83** [comment]
Other than a readme and sorting out those other comments you should be GTG AFAICT

**Comment by rothgar** [comment]
Thanks for the info. I have a few more changes and need to switch the default password to be randomly generated and print it out in the NOTES. I'll try to get to that today and should be able to finish up the chart.

**Comment by lachie83** [comment]
Okay great -- Just a word of warning RE: random value generation. https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#be-careful-with-generating-random-values

**Comment by rothgar** [comment]
Alright, this deployed for me on minikube and GKE.

I'd like to add options to not deploy postgres and redis but I'm still not exactly sure how to do that. Otherwise this deploys fine for me.

**Comment by rothgar** [comment]
Can someone go ahead and test this chart? I think it's ready to be merged.

**Comment by seanknox** [comment]
@k8s-bot ok to test

File: incubator/sentry/templates/hooks/user-create.job.yaml
Comment (seanknox) [inline incubator/sentry/templates/hooks/user-create.job.yaml]: Minor nit: use the same labels for all Kubernetes resources.

**Review by seanknox — COMMENTED** [review]
Great first chart! Just kicked off CI. Assuming it passes, last things to fix are 1) update name in Chart.yaml per Lachie and 2) use the same labels for all resources

File: incubator/sentry/templates/hooks/user-create.job.yaml
Comment (rothgar) [reply incubator/sentry/templates/hooks/user-create.job.yaml]: Good catch on the labels. I'll fix up some names and labels to be consistent across all resources.

**Comment by seanknox** [comment]
@rothgar Just noticed this chart is under `incubator/`, which is for charts that use alpha Kubernetes resources. We should do a better job of making that clear. In the meantime, you should move to `stable/`, though not blocking for this PR.

**Comment by rothgar** [comment]
Thanks for the clarification. I never did know what the difference was.

File: stable/sentry/Chart.yaml
Comment (seanknox) [inline stable/sentry/Chart.yaml]: Please update your name to your GitHub handle, so folks can easily ping you on future work on this chart.

**Review by seanknox — CHANGES_REQUESTED** [review]

**Comment by seanknox** [comment]
One last change and we should be gtg

**Comment by seanknox** [comment]
@k8s-bot ok to test

**Comment by seanknox** [comment]
Thanks @rothgar for your first chart!

**Comment by nhoughto** [comment]
I can't get this to run if i specify a release name in helm, only 

`helm install --wait stable/sentry`

works, if i specify `--name blah` then i get a timeout and failure. Looks like the user-create job fails connecting to the database. Is it expected that a release name can't be specified?

**Pull request merged**
